server/chat-plugins/abuse-monitor.ts

```diff
@@ -0,0 +1,719 @@
+/**
+ * Neural net chat'filters'.
+ * These are in a separate file so that they don't crash the other filters.
+ * (issues with globals, etc)
+ * We use Google's Perspective API to classify messages.
+ * @see https://perspectiveapi.com/
+ * by Mia.
+ * @author mia-pi-git
+ */
+
+import {FS, Utils, Net, ProcessManager, Repl} from '../../lib';
+import {Config} from '../config-loader';
+import {toID} from '../../sim/dex-data';
+
+const WHITELIST = ["mia"];
+// 20m. this is mostly here so we can use Monitor.slow()
+const PM_TIMEOUT = 20 * 60 * 1000;
+const ATTRIBUTES = {
+	"SEVERE_TOXICITY": {},
+	"TOXICITY": {},
+	"IDENTITY_ATTACK": {},
+	"INSULT": {},
+	"PROFANITY": {},
+	"THREAT": {},
+};
+
+export const cache: {
+	[roomid: string]: {[userid: string]: number} & {
+		staffNotified?: boolean,
+		claimed?: ID,
+	},
+} = global.Chat?.oldPlugins['abuse-monitor']?.cache || {};
+
+for (const k in cache) {
+	const entry = cache[k] as any;
+	if (entry.flags) delete entry.flags;
+	if (entry.users) {
+		cache[k] = {...entry.users, staffNotified: entry.notified};
+	}
+}
+
+const defaults: FilterSettings = {
+	threshold: 4,
+	minScore: 0.65,
+	specials: {
+		THREAT: {0.96: 'MAXIMUM'},
+		IDENTITY_ATTACK: {0.8: 2},
+		SEVERE_TOXICITY: {0.8: 2},
+	},
+};
+
+export const settings: FilterSettings = (() => {
+	try {
+		// accounting for data changes -
+		// make sure we do have the default data in case it's not in the stored data
+		return {...defaults, ...JSON.parse(FS('config/chat-plugins/nf.json').readSync())};
+	} catch (e: any) {
+		if (e.code !== "ENOENT") throw e;
+		return defaults;
+	}
+})();
+
+interface FilterSettings {
+	disabled?: boolean;
+	threshold: number;
+	minScore: number;
+	specials: {[k: string]: {[k: number]: number | "MAXIMUM"}};
+}
+
+export interface PerspectiveRequest {
+	languages: string[];
+	requestedAttributes: AnyObject;
+	comment: {text: string};
+}
+
+export interface PMResult {
+	score: number;
+	flags: string[];
+}
+
+function time() {
+	return Math.floor(Date.now() / 1000);
+}
+
+export class Limiter {
+	second = time();
+	countThisSecond = 0;
+	private counts = [] as number[];
+	private period: number;
+	max: number;
+	constructor(max: number, period: number) {
+		this.max = max;
+		this.period = period;
+	}
+	shouldRequest(depth = 0) {
+		const now = time();
+		if (this.second !== now) {
```

Comment:
The counter is wrong if multiple seconds pass between requests.

Reply:
Wrong how so? should we just have an interval that clears it once a second?

Reply:
Or, wait, no, I know what to do.

Reply:
Fixed.